include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp

```diff
@@ -409,9 +526,52 @@ namespace alpaka
                 return ::atomicExch(addr, value);
             }
         };
+        //! The GPU CUDA/HIPaccelerator atomic operation.
+        template<typename THierarchy>
+        struct AtomicOp<AtomicExch, AtomicUniformCudaHipBuiltIn, double, THierarchy>
+        {
+            __device__ static auto atomicOp(
+                AtomicUniformCudaHipBuiltIn const&,
+                double* const addr,
+                double const& value) -> double
+            {
+                unsigned long long int* address_as_ull(reinterpret_cast<unsigned long long int*>(addr));
+                unsigned long long int old(*address_as_ull);
+                unsigned long long int assumed;
+                do
+                {
+                    assumed = old;
+                    old = ::atomicCAS(
+                        address_as_ull,
+                        assumed,
+                        static_cast<unsigned long long>(__double_as_longlong(value)));
+                    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
+                } while(assumed != old);
+                return __longlong_as_double(static_cast<long long>(old));
+            }
+        };
```

Comment:
Atomic exchange is rather simple. Why do you need a CAS loop for that?

```suggestion
        template<typename THierarchy>
        struct AtomicOp<AtomicExch, AtomicUniformCudaHipBuiltIn, double, THierarchy>
        {
            __device__ static auto atomicOp(
                AtomicUniformCudaHipBuiltIn const&,
                double* const addr,
                double const& value) -> double
            {
                return __longlong_as_double(static_cast<long long>(
                    ::atomicExch(reinterpret_cast<unsigned long long*>(addr),
                    static_cast<unsigned long long>(__double_as_longlong(value))));
            }
        };
```